src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -426,5 +430,444 @@ public void EmptyRecord()
+  IL_0002:  newobj     ""C..ctor(int, int)""
+  IL_0007:  dup
+  IL_0008:  call       ""void System.Console.WriteLine(object)""
+  IL_000d:  ldc.i4.5
+  IL_000e:  stloc.0
+  IL_000f:  ldloc.0
+  IL_0010:  callvirt   ""C C.With(int)""
+  IL_0015:  call       ""void System.Console.WriteLine(object)""
+  IL_001a:  ret
+}");
+        }
+
+        [Fact]
+        public void WithExpr14()
+        {
+            var src = @"
+using System;
+data class C(int X, int Y)
+{
+    public C With(int X, int Y) => new C(X, Y);
+    public override string ToString() => X + "" "" + Y;
+    public static void Main()
+    {
+        var c = new C(0, 1);
+        Console.WriteLine(c);
+        c = c with { X = 5 };
+        Console.WriteLine(c);
+        c = c with { Y = 2 };
+        Console.WriteLine(c);
+    }
+}";
+            var verifier = CompileAndVerify(src, expectedOutput: @"0 1
+5 1
+5 2");
+            verifier.VerifyIL("C.Main", @"
+{
+  // Code size       55 (0x37)
+  .maxstack  3
+  .locals init (C V_0,
+                int V_1)
+  IL_0000:  ldc.i4.0
+  IL_0001:  ldc.i4.1
+  IL_0002:  newobj     ""C..ctor(int, int)""
+  IL_0007:  dup
+  IL_0008:  call       ""void System.Console.WriteLine(object)""
+  IL_000d:  stloc.0
+  IL_000e:  ldc.i4.5
+  IL_000f:  stloc.1
+  IL_0010:  ldloc.0
+  IL_0011:  ldloc.1
+  IL_0012:  ldloc.0
+  IL_0013:  callvirt   ""int C.Y.get""
+  IL_0018:  callvirt   ""C C.With(int, int)""
+  IL_001d:  dup
+  IL_001e:  call       ""void System.Console.WriteLine(object)""
+  IL_0023:  ldc.i4.2
+  IL_0024:  stloc.1
+  IL_0025:  dup
+  IL_0026:  callvirt   ""int C.X.get""
+  IL_002b:  ldloc.1
+  IL_002c:  callvirt   ""C C.With(int, int)""
+  IL_0031:  call       ""void System.Console.WriteLine(object)""
+  IL_0036:  ret
+}");
+        }
+
+        [Fact]
+        public void WithExpr15()
+        {
+            var src = @"
+data class C(int X, int Y)
+{
+    public C With(int X, int Y) => null;
+    public static void Main()
+    {
+        var c = new C(0, 0);
+        c = c with { = 5 };
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (8,22): error CS1525: Invalid expression term '='
+                //         c = c with { = 5 };
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "=").WithArguments("=").WithLocation(8, 22)
+            );
+        }
+
+        [Fact]
+        public void WithExpr16()
+        {
+            var src = @"
+data class C(int X, int Y)
+{
+    public C With(int X, int Y) => null;
+    public static void Main()
+    {
+        var c = new C(0, 0);
+        c = c with { X = };
+    }
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (8,26): error CS1525: Invalid expression term '}'
+                //         c = c with { X = };
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "}").WithArguments("}").WithLocation(8, 26)
+            );
+        }
```

Comment:
Please add some tests demonstrating the issues Neal and I brought up in other parts of the review. Additionally:

Tests with Y and X in opposite ordering, showing side effects are presevered.
Tests where the initializer expression needs to be converted to the parameter type, including user-defined conversions.
Tests where the return type of the with method is derived.
Tests where the instance is of an interface type.
Tests where the instance is of a struct type.